Spring for GraphQL and React (#1444)
oktadeveloper/okta-blog: _source/_posts/2023-07-27-spring-graphql-react.adoc

> ---

Reviewer: I love that you used AsciiDoc! Unfortunately, the Auth0 blog only supports Markdown. I can handle the translation for you if you don't have time.

Author: I migrated to Markdown

layout: blog_post
title: "How to build a GraphQL API with Spring for GraphQL"
author: jimena-garbarino
by: contractor
communities: [security,java,javascript]
description: "A step-by-step guide for building a secured GraphQL API with Spring Boot and Auth0 authentication on React"
tags: [java, javascript]
tweets:
- ""
- ""
- ""
image:
type: awareness
---
:page-liquid:
:toc: macro
:experimental:

> GraphQL is a query language for APIs and a runtime that allows the API consumer to get exactly the required information instead of the server exclusively controlling the contents of the response. While some REST API implementations require loading the references of a resource from multiple URLs, GraphQL APIs can follow references between related objects and return them in a single request. This step-by-step guide demonstrates how to build a GraphQL API with Spring Boot and Spring for GraphQL, for querying a sample dataset of related companies, persons, and properties, seeded to a Neo4j database. It also demonstrates how to build a React client with Next.js and MUI Datagrid to consume the API. Both the client and the server are secured with Auth0 authentication using Okta Spring Boot Starter for the server and Auth0 React SDK for the client.

Reviewer: I'd recommend adding a paragraph break before "This step-by-step..."

Author: Added

image::{% asset_path blog/spring-graphql-react/logos.png %}[alt=Spring, GraphQL and React,width=700,align=center]

[NOTE]
====
.This example was created with the following tools and services:
- https://docs.npmjs.com/downloading-and-installing-node-js-and-npm[Node.js v18.16.1]
- https://docs.npmjs.com/downloading-and-installing-node-js-and-npm[npm 9.5.1]
- https://jdk.java.net/java-se-ri/17[Java OpenJDK 17]
- https://docs.docker.com/desktop/[Docker 24.0.2]
- https://auth0.com/signup[Auth0 account]
- https://github.com/auth0/auth0-cli#installation[Auth0 CLI 1.0.0]
- https://httpie.io/[HTTPie 3.2.2]
- https://nextjs.org/[Next.js 13.4.19]
====

toc::[]

== Build a GraphQL API with Spring for GraphQL

The resource server is a Spring Boot web application that exposes a GraphQL API with Spring for GraphQL. The API allows querying a Neo4j database containing information about companies and their related owners and properties, using Spring Data Neo4j. The data was obtained from a Neo4j https://neo4j.com/graphgists/35a813ba-ea10-4165-9065-84f8802cbae8/[use case example].

Create the application with Spring Initializr and HTTPie:

[source,shell]
----
https start.spring.io/starter.zip \

>   bootVersion==3.1.2 \

Reviewer: ```suggestion
  bootVersion==3.1.3 \
```
Please update to the latest version of Spring Boot.

  language==java \
  packaging==jar \

>   javaVersion==17 \

Reviewer: These are all defaults. I don't think we need to include them, unless you want to be explicit.

Author: I would keep javaVersion explicit, and any other version setting.

  type==gradle-project \
  dependencies==data-neo4j,graphql,web \
  groupId==com.okta.developer \
  artifactId==spring-graphql  \
  name=="Spring Boot GraphQL API Application" \
  description=="Demo project of a Spring Boot GraphQL API" \
  packageName==com.okta.developer.demo > spring-graphql-api.zip
----

> Unzip the file and start editing the project. Define the GraphQL api with a schema file `schema.graphqls`, in the folder `src/main/resources/graphql/` with the following content:

Reviewer: ```suggestion
Unzip the file and start editing the project. Define the GraphQL API with a schema file named `schema.graphqls` in the `src/main/resources/graphql` directory:
```

        return new ReactiveNeo4jTransactionManager(driver, databaseNameProvider);
    }

}
----

Create the package `com.okta.developer.demo.controller` and the class `CompanyController` implementing the query endpoints matching the queries defined in the graphql schema:

.CompanyController.java
[source, java]
----
package com.okta.developer.demo.controller;

import com.okta.developer.demo.domain.Company;
import com.okta.developer.demo.repository.CompanyRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.graphql.data.method.annotation.Argument;
import org.springframework.graphql.data.method.annotation.QueryMapping;
import org.springframework.stereotype.Controller;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

@Controller
public class CompanyController {

    @Autowired
    private CompanyRepository companyRepository;

    @QueryMapping
    public Flux<Company> companyList(@Argument Long page) {
        return companyRepository.findAll().skip(page * 10).take(10);
    }

    @QueryMapping
    public Mono<Long> companyCount() {
        return companyRepository.count();
    }

}
----

> Create the `CompanyControllerTests` class for the web layer in the folder `src/main/test/java` under the package `com.okta.developer.demo.controller`:

Reviewer: ```suggestion
Create a `CompanyControllerTests` for the web layer in the folder `src/main/test/java` under the package `com.okta.developer.demo.controller`:
```

.CompanyControllerTests.java
[source, java]
----
package com.okta.developer.demo.controller;

import com.okta.developer.demo.domain.Company;
import com.okta.developer.demo.repository.CompanyRepository;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.graphql.GraphQlTest;
import org.springframework.boot.test.mock.mockito.MockBean;
import org.springframework.graphql.test.tester.GraphQlTester;
import reactor.core.publisher.Flux;

import java.time.LocalDate;

import static org.mockito.Mockito.when;

@GraphQlTest(CompanyController.class)
public class CompanyControllerTests {

    @Autowired
    private GraphQlTester graphQlTester;

    @MockBean
    private CompanyRepository companyRepository;

    @Test
    void shouldGetCompanies() {

> 

Reviewer: Remove extra line break.

Author: Removed

                .matchesJson("""
                    [{
                        "id": null,
                        "SIC": "1234",
                        "name": "Test Company",
                        "status": "active",
                        "category": "private",
                        "companyNumber": "12345678",
                        "countryOfOrigin": "UK"
                    }]
                """);
    }
}
----

Create the document file `companyList.graphql` containing the query definition for the test, in the folder `src/main/test/resources/graphql-test`:

.companyList.graphql
[source, graphql]
----
query companyList($page: Int) {
    companyList(page: $page) {
        id
        SIC
        name
        status
        category
        companyNumber
        countryOfOrigin
    }
}
----
Update the test configuration in `build.gradle` file, so passed tests are logged:

.build.gradle
[source, groovy]
----
tasks.named('test') {
  useJUnitPlatform()

  testLogging {
    // set options for log level LIFECYCLE
    events "failed", "passed"
  }
}
----

Run the test with:

[source, shell]
----
./gradlew clean test
----
You should see logs for the successful tests:

----
...

> SpringBootHraphQlApiApplicationTests > contextLoads() PASSED

Reviewer: This seems to have a typo in it. Does the class name in your repo have this same typo?

Author: Fixed

CompanyControllerTests > shouldGetCompany() PASSED
...
----

=== Add Neo4j seed data

Let's add Neo4j migrations dependency for the seed data insertion. Edit the file `build.gradle` and add:

.build.gradle
[source, groovy]
----
dependencies {
  ...

>   implementation 'eu.michael-simons.neo4j:neo4j-migrations-spring-boot-starter:2.4.0'

Reviewer: 2.5.3 is the latest version. https://central.sonatype.com/artifact/eu.michael-simons.neo4j/neo4j-migrations-spring-boot-starter/2.5.3

Author: About to test with latest version.

Author: The test went ok.

}
----

Create the folder `src/main/resources/neo4j/migrations` and the following migration files:

.V001__Constraint.cypher
[source, cypher]
----
{% raw %}
CREATE CONSTRAINT FOR (c:Company) REQUIRE c.companyNumber IS UNIQUE;
//Constraint for a node key is a Neo4j Enterprise feature only - run on an instance with enterprise
//CREATE CONSTRAINT ON (p:Person) ASSERT (p.birthMonth, p.birthYear, p.name) IS NODE KEY
CREATE CONSTRAINT FOR (p:Property) REQUIRE p.titleNumber IS UNIQUE;
{% endraw %}
----

.V002__Company.cypher
[source, cypher]
----
{% raw %}
LOAD CSV WITH HEADERS FROM "file:///PSCAmericans.csv" AS row
MERGE (c:Company {companyNumber: row.company_number})
RETURN COUNT(*);
{% endraw %}
----

.V003__Person.cypher
[source, cypher]
----
{% raw %}
LOAD CSV WITH HEADERS FROM "file:///PSCAmericans.csv" AS row
MERGE (p:Person {name: row.`data.name`, birthYear: row.`data.date_of_birth.year`, birthMonth: row.`data.date_of_birth.month`})
  ON CREATE SET p.nationality = row.`data.nationality`,
  p.countryOfResidence = row.`data.country_of_residence`

> // TODO: Address

Reviewer: Is this relevant? If so, does it get taken care of below?

Author: Removed, not addressed

c.SIC = row.`SICCode.SicText_1`,
c.countryOfOrigin = row.CountryOfOrigin,
c.status = row.CompanyStatus,
c.category = row.CompanyCategory;
{% endraw %}
----

.V006__Land.cypher
[source, cypher]
----
{% raw %}
LOAD CSV WITH HEADERS FROM "file:///LandOwnershipAmericans.csv" AS row
MATCH (c:Company {companyNumber: row.`Company Registration No. (1)`})
MERGE (p:Property {titleNumber: row.`Title Number`})
SET p.address = row.`Property Address`,
p.county  = row.County,
p.price   = toInteger(row.`Price Paid`),
p.district = row.District
MERGE (c)-[r:OWNS]->(p)
WITH row, c,r,p WHERE row.`Date Proprietor Added` IS NOT NULL
SET r.date = Date(Datetime({epochSeconds: apoc.date.parse(row.`Date Proprietor Added`,'s','dd-MM-yyyy')}));
CREATE INDEX FOR (c:Company) ON c.incorporationDate;
{% endraw %}
----

Update `application.properties` and add the following properties:

[source, properties]
----
spring.graphql.graphiql.enabled=true
spring.graphql.schema.introspection.enabled=true
org.neo4j.migrations.transaction-mode=PER_STATEMENT
spring.neo4j.uri=bolt://localhost:7687
spring.neo4j.authentication.username=neo4j
----

> Create a `.env` file int the project root to store the Neo4j credentials:

Reviewer: ```suggestion
Create a `.env` file in the project root to store the Neo4j credentials:
```

export interface CompanyData {
  id: string,
  name: string,
  category: string,
  companyNumber: string,
  SIC: string
}

export interface CompanyTableProps {
  rowCount: number,
  rows: CompanyData[],
  columns: GridColDef[],
  pagination: GridPaginationModel,
  onRowClick?: GridEventListener<"rowClick">
  onPageChange?: (pagination: GridPaginationModel) => void,

}

const CompanyTable = (props: CompanyTableProps) => {

  return (
    <>
      <DataGrid
        rowCount={props.rowCount}
        rows={props.rows}
        columns={props.columns}
        pageSizeOptions={[props.pagination.pageSize ]}
        initialState={{
          pagination: {
            paginationModel: { page: props.pagination.page, pageSize: props.pagination.pageSize },
          },
        }}
        density="compact"
        disableColumnMenu={true}
        disableRowSelectionOnClick={true}
        disableColumnFilter={true}
        disableDensitySelector={true}
        paginationMode="server"
        onRowClick={props.onRowClick}
        onPaginationModelChange={props.onPageChange}
      />
    </>
  );
};

export default CompanyTable;
{% endraw %}
----

> Create a `Loader.tsx` component in the folder `src/components/loader` with the following content:

Reviewer: ```suggestion
Create a `Loader.tsx` component in the folder `src/components/loader` with the following code:
```

.Loader.tsx
[source, tsx]
----
{% raw %}
import { Box, CircularProgress, Skeleton } from "@mui/material";

const Loader = () => {
  return (
    <Box sx={{ display: 'flex', justifyContent: 'center', alignItems: 'center', height: 200 }}>
      <CircularProgress />
    </Box>
  );
}

export default Loader;
{% endraw %}
----

> Add the file `src/components/company/CompanyTableContainer.tsx` with the following content:

Reviewer: ```suggestion
Add the file `src/components/company/CompanyTableContainer.tsx`:
```

  { field: "SIC", headerName: "SIC", width: 400, sortable: false },
];

const CompanyTableContainer = (props: CompanyTableProperties) => {
  const router = useRouter();
  const searchParams = useSearchParams()!;
  const pathName = usePathname();
  const page = props.page ? props.page : 1;

  const [ dataList, loadingList, errorList ] = useAsync(() => CompanyApi.getCompanyList({ page: page - 1 }), {}, [page]);
  const [ dataCount ] = useAsync(() => CompanyApi.getCompanyCount(), {}, []);

  const onPageChange = (pagination: GridPaginationModel) => {
    const params = new URLSearchParams(searchParams.toString());
    const page = pagination.page + 1;
    params.set("page", page.toString());
    router.push(pathName + "?" + params.toString());
  };

  return (
    <>
      {loadingList && <Loader/>}
      {errorList && <div>Error</div>}

      {!loadingList && dataList && (
        <CompanyTable
          pagination={{ page: page - 1, pageSize: 10 }}
          rowCount={dataCount}
          rows={dataList}
          columns={columns}
          onPageChange={onPageChange}
        ></CompanyTable>
      )}
    </>
  );
};

export default CompanyTableContainer;
{% endraw %}
----

> Add the file `src/app/HomePage.tsx` with the following content:

Reviewer: ```suggestion
Add `src/app/HomePage.tsx` for the homepage:
```

.HomePage.tsx
[source, tsx]
----
{% raw %}
"use client";

import CompanyTableContainer from "@/components/company/CompanyTableContainer";
import { Box, Typography } from "@mui/material";
import { useSearchParams } from "next/navigation";

const HomePage = () => {
  const searchParams = useSearchParams();
  const page = searchParams.get("page")
    ? parseInt(searchParams.get("page") as string)
    : 1;

  return (
    <>
      <Box>
        <Typography variant="h4" component="h1">
          Companies
        </Typography>
      </Box>
      <Box mt={2}>
        <CompanyTableContainer page={page}></CompanyTableContainer>
      </Box>
    </>
  );
};

export default HomePage;
{% endraw %}
----

> Replace the contents of `src/app/page.tsx` and set:

Reviewer: ```suggestion
Replace the contents of `src/app/page.tsx` and change it to render the `HomePage` component:
```

.app/page.tsx
[source, tsx]
----
{% raw %}
import HomePage from "./HomePage";

const Page = () => {
  return (
    <HomePage></HomePage>
  );
}

export default Page;
{% endraw %}
----

> Add a component defining the page width, for using it in the root layout. Create the file `src/layout/WideLayout.tsx` with the following content:

Reviewer: ```suggestion
Add a component defining the page width, for using it in the root layout. Create `src/layout/WideLayout.tsx` with the following content:
```

.WideLayout.tsx
[source, tsx]
----
{% raw %}
"use client";

import { Container, ThemeProvider, createTheme } from "@mui/material";

const theme = createTheme({
  typography: {
    fontFamily: "inherit",
  },
});

const WideLayout = (props: { children: React.ReactNode }) => {
  return (
    <ThemeProvider theme={theme}>
      <Container maxWidth="lg" sx={{ mt: 4 }}>
        {props.children}
      </Container>
    </ThemeProvider>
  );
};

export default WideLayout;
{% endraw %}
----

With the implementation above, the page content will be wrapped in a `ThemeProvider` component, so https://github.com/vercel/next.js/discussions/45433[MUI child components inherit the font family] from the root layout.

> Update the contents of `src/app/layout.tsx` and set:

Reviewer: ```suggestion
Update the contents of `src/app/layout.tsx` to be:
```

[source, shell]
----
auth0 apps create \
  --name "GraphQL Server" \
  --description "Spring Boot GraphQL Resource Server" \
  --type regular \
  --callbacks http://localhost:8080/login/oauth2/code/okta \
  --logout-urls http://localhost:8080 \
  --reveal-secrets
----

Once the app is created, you will see the OIDC app's configuration:

----
=== dev-avup2laz.us.auth0.com application created

  CLIENT ID            ***
  NAME                 GraphQL Server
  DESCRIPTION          Spring Boot GraphQL Resource Server
  TYPE                 Regular Web Application
  CLIENT SECRET        ***
  CALLBACKS            http://localhost:8080/login/oauth2/code/okta
  ALLOWED LOGOUT URLS  http://localhost:8080
  ALLOWED ORIGINS
  ALLOWED WEB ORIGINS
  TOKEN ENDPOINT AUTH
  GRANTS               implicit, authorization_code, refresh_token, client_credentials

 ▸    Quickstarts: https://auth0.com/docs/quickstart/webapp
 ▸    Hint: Emulate this app's login flow by running `auth0 test login ***`
 ▸    Hint: Consider running `auth0 quickstarts download ***`
----

Add the `okta-spring-boot-starter` dependency to the `build.gradle` file:
----
dependencies {
    ...

>   implementation 'com.okta.spring:okta-spring-boot-starter:3.0.4'

Reviewer: Please indent ... and implementation so it matches the build.gradle, which I believe is 4 spaces. At the very least, these lines should left-align.

Author: Fixed indentation

>   implementation 'com.okta.spring:okta-spring-boot-starter:3.0.4'

Reviewer: ```suggestion
  implementation 'com.okta.spring:okta-spring-boot-starter:3.0.5'
```

    ...
}
----

> Set the clientId, issuer and audience for OAuth2 in the `application.properties` file:

Reviewer: ```suggestion
Set the clientId, issuer, and audience for OAuth2 in the `application.properties` file:
```

Reviewer: Should it be OAuth2 or Auth0? I think it should be Auth0.

.application.properties
[source, properties]
----
okta.oauth2.issuer=https://{yourAuth0Domain}/
okta.oauth2.client-id={clientId}
okta.oauth2.audience=https://{yourAuth0Domain}/api/v2/
----

> Add the clientSecret to the `.env` file:

Reviewer: ```suggestion
Add the client secret to the `.env` file:
```

..env
[source, shell]
----
export OKTA_OAUTH2_CLIENT_SECRET={clientSecret}
----

Add the following factory method to the class `SpringBootGraphQLApiConfig`, for requiring a bearer token for all requests:

.SpringBootGraphQLApiConfig.java
[source, java]
----
    ...
    @Bean
    public SecurityFilterChain configure(HttpSecurity http) throws Exception {
        http.oauth2ResourceServer(oauth2ResourceServer -> oauth2ResourceServer.jwt(withDefaults()));
        return http.build();
    }
    ...

> ----

Reviewer: I believe this is already enabled by the Okta Spring Boot starter. I don't think it's necessary.

https://github.com/okta/okta-spring-boot/blob/master/oauth2/src/main/java/com/okta/spring/boot/oauth/OktaOAuth2AutoConfig.java#L88

Author: If I remove it, it seems the cors configuration at GraphQLConfig does not make effect, and requests from the react client fail. Not sure why is this happening.

Reviewer: If that's the case, maybe add a note about why it's necessary.

Author: Tested again, and added the note.

Again, in the root folder, run the API server with:
[source, shell]
----
source .env && ./gradlew bootRun
----

With HTTPie, send a request to the API server using a bearer access token:

[source, shell]
----
ACCESS_TOKEN={auth0AccessToken}
----

[source, shell]
----
echo -E '{"query":"{\n    companyList(page: 20) {\n        id\n        SIC\n        name\n        status\n        category\n        companyNumber\n        countryOfOrigin\n    }\n}"}' | \
  http -A bearer -a $ACCESS_TOKEN POST http://localhost:8080/graphql
----

[NOTE]
====
Follow https://auth0.com/docs/secure/tokens/access-tokens/get-management-api-access-tokens-for-testing[these instructions] to get an access token for the Auth0 Management API, which can be used for testing the server API.
====

=== Add Auth0 Login to the React client

> When using Auth0 as the identity provider, you can configure the Universal Login Page for a quick integration, without having to build the login forms. First, create an SPA application using the auth0 CLI:

Reviewer: ```suggestion
When using Auth0 as the identity provider, you can configure the Universal Login Page for a quick integration without having to build the login forms. First, create a SPA application using the Auth0 CLI:
```

[source, shell]
----
auth0 apps create \
  --name "React client for GraphQL" \
  --description "SPA React client for a Spring GraphQL API" \
  --type spa \
  --callbacks http://localhost:3000/callback \
  --logout-urls http://localhost:3000 \
  --reveal-secrets
----

> Copy the auth0 domain and the client Id, and update the `src/.env.local` with the following properties (add the new variables to the example file too):

Reviewer: ```suggestion
Copy the Auth0 domain and the client ID, and update the `src/.env.local` with the following properties (add the new variables to the example file, too):
```

.Auth0ProviderWithNavigate.tsx
[source, tsx]
----
{% raw %}
import { AppState, Auth0Provider } from "@auth0/auth0-react";
import { useRouter } from "next/navigation";
import React from "react";

const Auth0ProviderWithNavigate = (props: { children: React.ReactNode }) => {
  const router = useRouter();

  const domain = process.env.NEXT_PUBLIC_AUTH0_DOMAIN || "";
  const clientId = process.env.NEXT_PUBLIC_AUTH0_CLIENT_ID || "";
  const redirectUri = process.env.NEXT_PUBLIC_AUTH0_CALLBACK_URL || "";
  const audience = process.env.NEXT_PUBLIC_AUTH0_AUDIENCE || "";

  const onRedirectCallback = (appState?: AppState) => {
    router.push(appState?.returnTo || window.location.pathname);
  };

  if (!(domain && clientId && redirectUri)) {
    return null;
  }

  return (
    <Auth0Provider
      domain={domain}
      clientId={clientId}
      authorizationParams={{
        audience: audience,
        redirect_uri: redirectUri,
      }}
      useRefreshTokens={true}
      onRedirectCallback={onRedirectCallback}
    >
      <>{props.children}</>
    </Auth0Provider>
  );
};

export default Auth0ProviderWithNavigate;
{% endraw %}
----

> The component `Auth0ProviderWithNavigate` wraps the children component with `Auth0Provider`, the provider of the auth0 context, remembering the requested URL for redirection after login.

Reviewer: ```suggestion
The component `Auth0ProviderWithNavigate` wraps the children component with `Auth0Provider`, the provider of the Auth0 context, remembering the requested URL for redirection after login.
```

import { useEffect } from "react";
import Loader from "../loader/Loader";

const AuthenticationGuard = (props: { children: React.ReactNode }) => {
  const { isLoading, isAuthenticated, error, loginWithRedirect } = useAuth0();

  useEffect(() => {
    if (!isAuthenticated && !isLoading) {
      loginWithRedirect({
        appState: { returnTo: window.location.href },
      });
    }
  }, [isAuthenticated, isLoading, loginWithRedirect]);

  if (isLoading) {
    return <Loader />;
  }
  if (error) {
    return <div>Oops... {error.message}</div>;
  }
  return <>{isAuthenticated && props.children}</>;
};

export default AuthenticationGuard;
----

The `AuthenticationGuard` component will be used to protect pages that require authentication, redirecting to the Auth0 universal login. Protect the index page by wrapping its content in the `AuthenticationGuard` component:

.app/page.tsx
[source, tsx]
----
{% raw %}
import AuthenticationGuard from "@/components/authentication/AuthenticationGuard";
import HomePage from "./HomePage";

const Page = () => {
  return (
    <AuthenticationGuard>
      <HomePage></HomePage>
    </AuthenticationGuard>
  );
};

export default Page;
{% endraw %}
----

> === Call the API server with an accessToken

Reviewer: Most instances of "accessToken" should probably be "access token", unless it's used as a variable in code.

Author: Fixed

> Add th file `src/services/auth.tsx` with the following code:

Reviewer: ```suggestion
Add the file `src/services/auth.tsx` with the following code:
```

        headers: {
          ...config.headers,
          Authorization: `Bearer ${accessToken}`,
        },
      };
    },
    function (error) {
      console.log("request interceptor error", error);
      return Promise.reject(error);
    }
  );
};
----

Add the file `src/hooks/useAccessToken.tsx` with the following content:

.useAccessToken.tsx
[source, tsx]
----
import { setInterceptors } from "@/services/auth";
import { useAuth0 } from "@auth0/auth0-react";
import { useCallback, useState } from "react";

export const useAccessToken = () => {
  const { isAuthenticated, getAccessTokenSilently } = useAuth0();
  const [accessToken, setAccessToken] = useState("");

  const saveAccessToken = useCallback(async () => {
    if (isAuthenticated) {
      try {
        const tokenValue = await getAccessTokenSilently();
        if (accessToken !== tokenValue) {
          setInterceptors(tokenValue);
          setAccessToken(tokenValue);
        }
      } catch (err) {
        // Inactivity timeout
        console.log("getAccessTokenSilently error", err);
      }
    }
  }, [getAccessTokenSilently, isAuthenticated, accessToken]);

  return {
    saveAccessToken,
  };
};
----

> The hook will call Auth0 `getAccessTokenSilently` and will trigger a token refresh if the accessToken is expired. Then it will update axios interceptors to set the updated bearer token value in the request headers.

Reviewer: ```suggestion
The hook will call Auth0 `getAccessTokenSilently()` and will trigger a token refresh if the access token is expired. Then it will update axios interceptors to set the updated bearer token value in the request headers.
```

Create the `useAsyncWithToken` hook:

.useAsyncWithToken.tsx
[source, tsx]
----
import { useAccessToken } from "./useAccessToken";
import { useAsync } from "react-use-custom-hooks";

export const useAsyncWithToken = <T, P, E = string>(
  asyncOperation: () => Promise<T>, deps: any[]
) => {
  const { saveAccessToken } = useAccessToken();

  // Add saveAccessToken to the dependency list
  const [ data, loading, error ] = useAsync(async () => {
    await saveAccessToken();
    return asyncOperation();
  }, {},  deps);

  return {
    data,
    loading,
    error
  };
};
----

Update the calls in the `CompanyTableContainer` component to use the `useAsyncWithToken` hook:

.CompanyTableContainer.tsx
[source, tsx]
----
  ...
  const { data: dataList, loading: loadingList, error: errorList } = useAsyncWithToken(() => CompanyApi.getCompanyList({ page: page - 1 }), [page]);
  const { data: dataCount } = useAsyncWithToken(() => CompanyApi.getCompanyCount(), []);

>   ...

Reviewer: If there's no surrounding code, you can left-align all lines in this block.

Author: Aligned

    () => CompanyApi.getCompanyCount(),
    []
  );

  const onPageChange = (pagination: GridPaginationModel) => {
    const params = new URLSearchParams(searchParams.toString());
    const page = pagination.page + 1;
    params.set("page", page.toString());
    router.push(pathName + "?" + params.toString());
  };

  const companyData = dataList?.map((company: CompanyDTO) => {
    return {
      id: company.id,
      name: company.name,
      category: company.category,
      companyNumber: company.companyNumber,
      SIC: company.SIC,
      status: company.status,
      owner: company.controlledBy.map((person) => person.name).join(", "),
    }
  });

  return (
    <>
      {loadingList && <Loader />}
      {errorList && <div>Error</div>}

      {!loadingList && dataList && (
        <CompanyTable
          pagination={{ page: page - 1, pageSize: 10 }}
          rowCount={dataCount}
          rows={companyData}
          columns={columns}
          onPageChange={onPageChange}
        ></CompanyTable>
      )}
    </>
  );
};

export default CompanyTableContainer;
{% endraw %}
----

Give it a try, and notice how simple was the update.

== Learn More

> I hope you enjoyed this tutorial, and found this example useful. As you can see, not much work would be required to consume more company data from the GraphQL server, just a query update in the client. Also, the Auth0 Universal Login and Auth0 React SDK provide an efficient way to secure your React applications, following security best practices. Check the Auth0 documentation for adding https://developer.auth0.com/resources/guides/spa/react/basic-authentication#add-user-sign-up-to-react[sign-up] and https://developer.auth0.com/resources/guides/spa/react/basic-authentication#add-user-logout-to-react[logout] functionality to your React application. You can find all the code for this example in the https://github.com/indiepopart/spring-graphql-react[Github repository].

Reviewer: It looks like this needs to be broken into paragraphs.

Author: Added a break

> Keep in touch! If you have questions about this post, please ask them in the comments below. And follow us! We're https://twitter.com/oktadev[@oktadev on Twitter], https://youtube.com/c/oktadev[@oktadev on YouTube], and frequently post to our https://www.linkedin.com/company/oktadev/[LinkedIn page].

Reviewer: Todo: add links to other Spring Boot / Java posts, as well as a CTA for the newsletter.

```
You can also sign up for our [newsletter](https://a0.to/nl-signup/java) to stay updated on everything Identity and Security.
```

Author: Added

I hope you enjoyed this tutorial, and found this example useful. As you can see, not much work would be required to consume more company data from the GraphQL server, just a query update in the client. Also, the Auth0 Universal Login and Auth0 React SDK provide an efficient way to secure your React applications, following security best practices. You can find all the code for this example in the https://github.com/indiepopart/spring-graphql-react[Github repository]. Check the Auth0 documentation for adding https://developer.auth0.com/resources/guides/spa/react/basic-authentication#add-user-sign-up-to-react[sign-up] and https://developer.auth0.com/resources/guides/spa/react/basic-authentication#add-user-logout-to-react[logout] functionality to your React application.

Check out some other fun tutorials for Spring Boot and React:

- https://auth0.com/blog/build-crud-spring-and-vue/[Build a Simple CRUD App with Spring Boot and Vue.js]
- https://auth0.com/blog/simple-crud-react-and-spring-boot/[Use React and Spring Boot to Build a Simple CRUD App]
- https://auth0.com/blog/full-stack-java-with-react-spring-boot-and-jhipster/[Full Stack Java with React, Spring Boot, and JHipster]

> Keep in touch! If you have questions about this post, please ask them in the comments below. And follow us! We're https://twitter.com/oktadev[@oktadev on Twitter], https://youtube.com/c/oktadev[@oktadev on YouTube], and frequently post to our https://www.linkedin.com/company/oktadev/[LinkedIn page]. You can also sign up for our   https://a0.to/nl-signup/java[newsletter] to stay updated on everything Identity and Security.

Reviewer: ```suggestion
Keep in touch! If you have questions about this post, please ask them in the comments below. And follow us! We're https://twitter.com/oktadev[@oktadev on Twitter], https://youtube.com/c/oktadev[@oktadev on YouTube], and frequently post to our https://www.linkedin.com/company/oktadev/[LinkedIn page]. You can also sign up for our https://a0.to/nl-signup/java[newsletter] to stay updated on everything Identity and Security.
```
